### PR TITLE
Added new chapter about GRUB authorization

### LINKED ISSUE #2115: Add documentation about GRUB authorization
## Problem description
I have already described the topic in the SUSE Administration Guide (see https://github.com/SUSE/doc-sle/pull/1164/files#diff-f47b6d53eaf93443eb5f106c1b5a83389597f6f4d3fcd961bfd827f97473a1f7R726) and now would like to update upstream KIWI documentation to include it as well.

### PR DESCRIPTION
**Pull request by tbazant**
Fixes #2115  .

Changes proposed in this pull request:
* introduces custom boot loader configuration to allow authorizing boot menu entries

### CODE REVIEW

File: doc/source/working_with_images/grub_authorization.rst [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     .. code:: bash
+
+        $ sudo grub2-mkconfig -o /boot/grub2/grub.cfg
+
+In the above example:
+
+* The GRUB menu has two entries, *Operational mode* and *Maintenance mode*.
+* If no user is specified, both boot menu entries are accessible, but
+  no one can access GRUB command line nor edit existing menu entries.
+* `admin` user can access GRUB command line and edit existing menu entries.
```
Comment (dcermak) [inline doc/source/working_with_images/grub_authorization.rst:69]:
```suggestion
* the `admin` user can access the GRUB command line and edit existing menu entries.
```

**Review by dcermak — CHANGES_REQUESTED** [review]
The text itself is very good, albeit I cannot attest its factual correctness, as I know far too little about grub.

However, this is a very general guide and imho is not fit for the context of kiwi, as it does not explain how you apply this when building a kiwi image. For instance, when do you invoke all these commands? How do you apply these settings to your image? Do you put some scripts into `config.sh` or `images.sh`? etc.

File: doc/source/working_with_images/grub_authorization.rst [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+   GRUB. Use the `grub2-mkpasswd-pbkdf2`:
+
+   .. code:: bash
+
+      $ sudo grub2-mkpasswd-pbkdf2
+      Password: ****
+      Reenter password: ****
+      PBKDF2 hash of your password is grub.pbkdf2.sha512.10000.9CA4611006F77A...
+
+2. Delete the content of the `/etc/grub.d/10_linux` file and save it.
```
Comment (davidcassany) [inline doc/source/working_with_images/grub_authorization.rst:27]: Ok, from a builder perspective you can't delete something that is not already generated. So the strategy should be to provide a KIWI project config that from a fresh build it produces an image with customized grub2.

So in KIWI terms this step should be part of `config.sh` script that is executed in chroot after installing all the packages (so in that stage there should already be a `/etc/grub.d/10_linux` file). Moreover, why do we need to save it? If this script is only the generator of the default menu entries we don't want we can simply remove it, don't we?

File: doc/source/working_with_images/grub_authorization.rst [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+   This prevents outputting the default GRUB menu entries.
+
+3. Edit the `/boot/grub2/custom.cfg` file and add custom
```
Comment (davidcassany) [inline doc/source/working_with_images/grub_authorization.rst:30]: This file should be provided as part of the overlay root in KIWI so it gets included into the image build to the expected path with the expected contents. So this step should explicitly tell it should be part of the KIWI's overlay root.

File: doc/source/working_with_images/grub_authorization.rst [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+      set superusers=admin
+      password admin ADMIN_PASSWORD
+      password maintainer MAINTAINER_PASSWORD
+
+      menuentry 'Operational mode' {
+        insmod ext2
+        set root=hd0,1
+        echo 'Loading Linux ...'
+        linux /boot/vmlinuz root=/dev/vda1 $GRUB_CMDLINE_LINUX_DEFAULT $GRUB_CMDLINE_LINUX mode=operation
```
Comment (davidcassany) [inline doc/source/working_with_images/grub_authorization.rst:44]: KIWI runs some patches over the generated grub.cfg to ensure it is EFI/BIOS bootable and to provide a proper `root=` kernel parameter. We need to check this template does not interfere over the grub2 configuration steps neither produce inconsistent data.

**Review by davidcassany — COMMENTED** [review]
I like the idea of documenting this topic, as it can be an example to showcase how grub2 configuration can be customized in KIWI context.

However this docs they look like the steps you would run on an already existing system. I am not sure they properly match image builder PoV. Have you verified that in a KIWI image build process or just applied it on a KIWI built image? We need a full step explanation starting from one of the kiwi-descriptions examples, or alternatively provide an example in kiwi-descriptions that actually runs this setup. See some comments in line.

The docs here I'd suggest them to encompass them in the scope of gurb2 configuration customization example and use grub2 authorization as nice example for it since we'd define new menu entries with highly customized menu entries.

**Comment by davidcassany** [comment]
@schaefi what's you opinion regarding this? I think this approach of setting the `/etc/grub.d/custom.cfg` is nice and should be compatible with KIWI's process. I was thinking on giving it a try to this procedure and check if our patches do not break it, I believe it can work and it would be a nice way to provide customized grub2 config. 

Obviously there are many ways to break the resulting image in that procedure but if it works without having to consider too many constraints I think this is a great example to document and ti should be the preferred way to customize grub2 as it is pretty compliant with a regular grub2 usage.

**Comment by schaefi** [comment]
> @schaefi what's you opinion regarding this? I think this approach of setting the `/etc/grub.d/custom.cfg` is nice and should be compatible with KIWI's process. I was thinking on giving it a try to this procedure and check if our patches do not break it, I believe it can work and it would be a nice way to provide customized grub2 config.

I agree and I remember that I already used this concept for a customer project at Ericson, see here:
https://build.opensuse.org/package/show/Virtualization:Appliances:Images/grub-custom-menu

So yes I'm also in favor of making use of this concept
 
> Obviously there are many ways to break the resulting image in that procedure but if it works without having to consider too many constraints I think this is a great example to document and ti should be the preferred way to customize grub2 as it is pretty compliant with a regular grub2 usage.

I agree and it aligns with the way we documented and promote custom dracut.conf.d/... extensions.
Thus +1 from my side

I have to admit I did not yet had time to read the proposed documentation of this pull request. I hope I get to it within the next days

**Comment by Conan-Kudo** [comment]
Why isn't this in upstream GRUB documentation?

**Comment by schaefi** [comment]
@tbazant Hi Thomas. We discussed the open pull request during the community meeting and thought about how to continue with the documentation you wrote. First of all thanks much for taking the effort to write this up and send it here. There is however one issue which prevents us from merging this in the kiwi space. The documentation is a valuable description how authorization in grub works and actually deserves to be a part of the grub documentation. In its current form it however does not fit to kiwi. The reason here is that the instructions how to configure authorization for grub are not connected to how you would do this in a kiwi image description. Like @davidcassany mentioned:

> However this docs they look like the steps you would run on an already existing system.

I have the same impression and thus we cannot add the docs in its current form.

My suggestion and also to value the work you have done is as follows:

1. We identified the grub upstream documentation does not provide the information you documented. Thus we would take the docs you wrote and put them into a gist.
2. We contact "our" grub developers and point them to the docs in the gist asking them to add this information to the general grub documentation
3. For using grub authorization in kiwi we need a new pull request which focuses on a procedure to do the needed changes in a grub setup as part of a kiwi image description. This information can point to the grub docs for details how authorization works.
4. Along with the proposed way to setup grub authorization in kiwi an integration test image should be provided to demonstrate that it works.

This in the end leads to a helpful documentation and example users of kiwi can benefit from.

Please let us know if you are fine with this procedure ? If so I can create the gist with your data. For the upstream grub update we can only hope for the people to take it. Regarding 3. this needs some background knowledge about kiwi and appliances. It would be good if you can share your thoughts about this part. Meaning are you interested to dig into kiwi a bit more or would you rather focus on the documentation only ? Providing an integration test image is something I can help with combined with the new doc pull request to tackle 3.

What do you think ?

Thanks much

**Comment by tbazant** [comment]
Hi, see below

Marcus Schäfer píše v Út 10. 05. 2022 v 07:28 -0700:
> @tbazant Hi Thomas. We discussed the open pull request during the
> community meeting and thought about how to continue with the
> documentation you wrote. First of all thanks much for taking the
> effort to write this up and send it here. There is however one issue
> which prevents us from merging this in the kiwi space. The
> documentation is a valuable description how authorization in grub
> works and actually deserves to be a part of the grub documentation.

you're welcome

> In its current form it however does not fit to kiwi. The reason here
> is that the instructions how to configure authorization for grub are
> not connected to how you would do this in a kiwi image description.
> Like @davidcassany mentioned:
> > However this docs they look like the steps you would run on an
> > already existing system.
> I have the same impression and thus we cannot add the docs in its
> current form.
> My suggestion and also to value the work you have done is as follows:
>    1. We identified the grub upstream documentation does not provide
> the information you documented. Thus we would take the docs you wrote
> and put them into a gist.
>    2. We contact "our" grub developers and point them to the docs in
> the gist asking them to add this information to the general grub
> documentation
>    3. For using grub authorization in kiwi we need a new pull request
> which focuses on a procedure to do the needed changes in a grub setup
> as part of a kiwi image description. This information can point to
> the grub docs for details how authorization works.
>    4. Along with the proposed way to setup grub authorization in kiwi
> an integration test image should be provided to demonstrate that it
> works.
> This in the end leads to a helpful documentation and example users of
> kiwi can benefit from.
> Please let us know if you are fine with this procedure ? If so I can
> create the gist with your data.

i'm totally fine with what you suggested

> For the upstream grub update we can only hope for the people to take
> it. Regarding 3. this needs some background knowledge about kiwi and
> appliances. It would be good if you can share your thoughts about
> this part. Meaning are you interested to dig into kiwi a bit more or
> would you rather focus on the documentation only ? Providing an
> integration test image is something I can help with combined with the
> new doc pull request to tackle 3.
> What do you think ?

take my current PR as a source of information and do whatever you like
to do with it please. i'm not going to dig deeper into KIWI, nor create
another PR. Time does not permit me to do so. At SUSE, i'm assigned to
other projects and am currently busy with SLE 15 SP4 release.

Sorry that i cannot do more for you at this time.

cheers

Tomáš

**Comment by schaefi** [comment]
> I'm totally fine with what you suggested

ok thanks. I have moved your documentation into a public gist here:

* https://gist.github.com/schaefi/5ade26d4d4f6007fc6cf17b3bea01e67

@Conan-Kudo would you like to contact the grub developers you know and send them this information ? I only know mchang@suse.de and can write him an e-mail but I think you are known in a broader scope than I am :) In any case would be nice to get CC'ed. Thanks

**Comment by schaefi** [comment]
> I'm not going to dig deeper into KIWI, nor create another PR.

ok that's fair enough, thanks for the quick response. In this case we will wait for the response of the grub dev team and accrding to their reaction we plan for next steps. I'll keep this PR open for some time and will then close it on behalf of the existing issue description. Thanks

**Comment by schaefi** [comment]
I have sent out the information about the documentation for grub to mchang from suse who is one of the people I know working on grub upstream. So far no response

Closing this one here. Thanks for your effort

**Pull request closed without merging**